### PR TITLE
Fix broken SCM tree

### PR DESCRIPTION
**Pull request by westbury**
Fixes https://github.com/eclipse-theia/theia/issues/8004

#### What it does
The code that generated the tree in the Source Control view assumed that files came back from the SCM sorted, or at least with files in the same folder kept together.  Git mostly does, but not always.  So this PR sorts the files before building the tree.

#### How to test
1. Checkout branch theia-7577
2. Amend all the commits in the branch, back to and including the initial commit.
3. See the tree under 'Staged Changes'.  This tree is correct.
4. Unstage all.  You will now see the tree incorrect (without this PR), correct with this PR.

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

**Comment by akosyakov** [comment]
@vince-fugnitto Could you help reviewing it? or find a reviewer?

**Comment by lmcbout** [comment]
Applying "How to Test", I don't see any differences between this patch and the master branch commit. It seems to be the same. What exactly should not be the same ? Is there something specific in the tree ?

**Comment by westbury** [comment]
> I don't see any differences between this patch and the master branch commit.

Did you unstage the changes?

The problem is caused because 'git status' returns the untracked files last.  So if a folder contains both tracked and untracked files then the folder appears twice in the tree (except it doesn't because it will have the same id in both instances).  Have a look at the screen shot in https://github.com/eclipse-theia/theia/issues/8004.

The problem occurs for both Theia's git package and vs code's builtin.

**Comment by lmcbout** [comment]
I followed the 
**How to test**
Checkout branch theia-7577
Amend all the commits in the branch, back to and including the initial commit.
``` 
(git reset  --soft c20fed2b5045392e113071176a284addfbde84fa to get all files)
```
See the tree under 'Staged Changes'. This tree is correct.
--> See the tree with all items marked with "A"
Unstage all. 
--> See the tree with all files maked "U" 
--> Tree is moved from 'Staged Changes' to `Changes` 

Samething with / without this PR
Note: The screenshot is without your PR
Tested on Ubuntu 18.04
Is it specific to a platform ?

GitStageUnstage

**Comment by westbury** [comment]
@lmcbout Sorry, I meant the initial commit in the branch, not the initial commit in the repository.  That would be resetting to daede30361f6381614be07c9d9405a9a1b2944bb.  I just checked out the branch then pressed the 'Amend' button three times (to amend the two merge commits plus the first commit in the branch).

You just need to get both a modified file and an untracked file into the CHANGES (unstaged) section.  These two files also must be in the same folder and not at the root of the repository.

File: packages/scm/src/browser/scm-tree-model.ts [hunk 1]
```diff
@@ -207,19 +211,19 @@ export class ScmTreeModel extends TreeModelImpl {
                 if (folderEnd - folderStart < nestingThreshold) {
```
Comment (tsmaeder) [inline packages/scm/src/browser/scm-tree-model.ts:211]: To me, the name "nestingThreshold" seems to indicate the depth of the tree. I don't see how the difference between two different indices in the list of resources is related to that. Shouldn't this expression involve the `level` somehow?

File: packages/scm/src/browser/scm-tree-model.ts [hunk 2]
```diff
@@ -207,19 +211,19 @@ export class ScmTreeModel extends TreeModelImpl {
...
                     let thisLevel = level + 1;
                     while (thisLevel < firstFileParts.length - 1 && thisLevel < lastFileParts.length - 1 && firstFileParts[thisLevel] === lastFileParts[thisLevel]) {
                         thisLevel++;
                     }
                     const nodeRelativePath = firstFileParts.slice(level, thisLevel).join('/');
-                    result.push(this.toFileChangeFolderNode(resources, folderStart, folderEnd, thisLevel, nodeRelativePath, parent));
+                    result.push(this.toFileChangeFolderNode(sortedResources, folderStart, folderEnd, thisLevel, nodeRelativePath, parent));
                 }
                 folderStart = folderEnd;
             }
```
Comment (tsmaeder) [inline packages/scm/src/browser/scm-tree-model.ts:229]: I think we can get rid of `result.sort(...)` if we keep separate result arrays for file nodes and folder nodes and concatenate them at the end. With your change, the list of resources is already sorted and as far as I can see, the sorting will be correct under the algorithm.

**Comment by tsmaeder** [comment]
@westbury is this algorithm time-critical? I have an idea to make it O(n), just wondering if it would be worth giving it a try.

**Comment by westbury** [comment]
Thanks for your reviews.

@tsmaeder You are correct that the files and folders can be pushed to two lists and then appended, so O(n) rather than O(n log n).  I would rather take the time to do a fuller performance investigation as we have a ticket against Mbed Studio complaining about the 1000 limit on the changes status.  So I'll merge this fix for now.

**Pull request merged**